grpc/internal/meta/session.go

```diff
@@ -0,0 +1,94 @@
+package meta
+
+import (
+	"context"
+	"sync"
+
+	"github.com/keratin/authn-server/app"
+	"github.com/keratin/authn-server/app/models"
+	"github.com/keratin/authn-server/app/tokens/sessions"
+	pkgerrors "github.com/pkg/errors"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+type sessionKey int
+type accountIDKey int
+
+func SessionInterceptor(app *app.App) grpc.UnaryServerInterceptor {
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+		var session *sessions.Claims
+		var parseOnce sync.Once
+		parse := func() *sessions.Claims {
+			parseOnce.Do(func() {
+				md, ok := metadata.FromIncomingContext(ctx)
+				if !ok {
+					return
+				}
+				cookies := md.Get(app.Config.SessionCookieName)
+				if len(cookies) == 0 {
+					return
+				}
+				var err error
+				session, err = sessions.Parse(cookies[0], app.Config)
+				if err != nil {
+					app.Reporter.ReportError(pkgerrors.Wrap(err, "Parse"))
+					return
+				}
+			})
+			return session
+		}
+
+		var accountID int
+		var lookupOnce sync.Once
+		lookup := func() int {
+			lookupOnce.Do(func() {
+				var err error
+				session := parse()
+				if session == nil {
+					return
+				}
+
+				accountID, err = app.RefreshTokenStore.Find(models.RefreshToken(session.Subject))
+				if err != nil {
+					app.Reporter.ReportError(pkgerrors.Wrap(err, "Find"))
```

Comment:
Using `LogReporter` implementation of `app.Reporter.ReportRequestError` as example, perhaps I can add `ReportGRPCError(err error, info *grpc.UnaryServerInfo, req interface{})` to the `ErrorReporter` interface, and the LogReporter implementation of it can be something like this:

```go
func (r *LogReporter) ReportGRPCError(err error, info *grpc.UnaryServerInfo, req interface{}) {
	log.Printf("[%v][%T %v] [%T] %v\n", time.Now(), info.Server, info.FullMethod, req, err)
}
```

This will log which server implementation had the error and which method is the target, in addition to the error itself.

How does this sound? Printing the `req` can be harmful as it can contain passwords, so only the type is reasonable. The metadata stored in the context can also be sensitive because it can have the Basic Auth data.

Reply:
Yes, that looks good. I defer to your judgement on what details are useful for debugging a GRPC error. Thanks for making sure to not print passwords or sessions.

The SentryReporter and AirbrakeReporter implementations can probably satisfy the new interface by passing the error along to `ReportError`, without any extra context.

Reply:
I've added the method. I added context information for Airbrake and Sentry because I think it'll be confusing without them given there are multiple services in the application (although I could be completely wrong because I'm not familiar with Sentry and/or Airbrake).

I'll play with a Sentry docker container to see how the current implementation of reporting plays out with the overall implementation.